numpy/linalg/_linalg.py

```diff
@@ -2067,12 +2072,12 @@ def matrix_rank(A, tol=None, hermitian=False):
     if A.ndim < 2:
         return int(not all(A == 0))
     S = svd(A, compute_uv=False, hermitian=hermitian)
+    if rtol is not None and tol is not None:
+        raise ValueError("`tol` and `rtol` can't be both set.")
+    if rtol is None:
+        rtol = max(A.shape[-2:]) * finfo(S.dtype).eps
     if tol is None:
-        tol = (
-            S.max(axis=-1, keepdims=True) *
-            max(A.shape[-2:]) *
-            finfo(S.dtype).eps
-        )
+        tol = S.max(axis=-1, keepdims=True) * rtol
```

Comment:
I think this logic is incorrect. `tol` should also have a newaxis appended in the end in this case too (i.e., the `else` below should be removed and the `tol = asarray(tol)[..., newaxis]` line should be run unconditionally). Right now we have:

```py
>>> import numpy as np
>>> x = np.zeros((4, 3, 2))
>>> rtol = np.zeros((4,))
>>> np.linalg.matrix_rank(x, tol=rtol)
array([0, 0, 0, 0])
>>> np.linalg.matrix_rank(x, rtol=rtol)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/numpy/numpy/linalg/_linalg.py", line 2083, in matrix_rank
    return count_nonzero(S > tol, axis=-1)
                         ^^^^^^^
ValueError: operands could not be broadcast together with shapes (4,2) (4,4)
```

The broadcasting for `rtol` should be the same as for `tol`. It should broadcast against the stack shape (`x.shape[:-2]`).

Reply:
That's right - Here's a fix for it: https://github.com/numpy/numpy/pull/25877